File: react-native/__tests__/quirks.test.ts
import { HingeAngleRange, HingeResolution } from '../src/model/hinge';
import {
  ASSUMED_QUIRK,
  lookupQuirk,
  normalizeAngle,
  registerQuirk,
  resetQuirks,
} from '../src/model/quirks';

afterEach(resetQuirks);

describe('normalisation', () => {
  it('passes through a standard 0-180 device', () => {
    expect(normalizeAngle(90, ASSUMED_QUIRK)).toBe(90);
    expect(normalizeAngle(0, ASSUMED_QUIRK)).toBe(0);
    expect(normalizeAngle(180, ASSUMED_QUIRK)).toBe(180);
  });

  it('null in, null out — a missing sensor never becomes zero', () => {
    expect(normalizeAngle(null, ASSUMED_QUIRK)).toBeNull();
    expect(normalizeAngle(undefined, ASSUMED_QUIRK)).toBeNull();
  });

  it('mirrors a 0-360 device past the flat point', () => {
    const quirk = { ...ASSUMED_QUIRK, range: HingeAngleRange.ZeroTo360 };
    expect(normalizeAngle(180, quirk)).toBe(180);
    expect(normalizeAngle(270, quirk)).toBe(90);
    expect(normalizeAngle(360, quirk)).toBe(0);
  });

  it('inverts a device that reports 180 closed', () => {
    const quirk = { ...ASSUMED_QUIRK, inverted: true };
    expect(normalizeAngle(0, quirk)).toBe(180);
    expect(normalizeAngle(180, quirk)).toBe(0);
  });

  it('applies an offset before normalising', () => {
    expect(normalizeAngle(85, { ...ASSUMED_QUIRK, offset: 5 })).toBe(90);
  });

  it('clamps out-of-range readings', () => {
    expect(normalizeAngle(-10, ASSUMED_QUIRK)).toBe(0);
    expect(normalizeAngle(200, ASSUMED_QUIRK)).toBe(180);
  });
});

describe('lookup', () => {
  it('falls back to the assumed convention for unknown devices', () => {
    expect(lookupQuirk('nothing', 'phone-3').range).toBe(
      HingeAngleRange.ZeroTo180,
    );
  });

  it('an unknown device does not claim to be continuous', () => {
    // Assuming a smooth sweep and being wrong produces an effect that
    // visibly snaps, which is worse than not offering the effect.
    expect(lookupQuirk('nothing', 'phone-3').resolution).toBe(
      HingeResolution.Unknown,
    );
  });

  it('finds shipped entries without registration', () => {
    // Measured on hardware: the Z Flip 5 reports exactly three values.
    const quirk = lookupQuirk('samsung', 'SM-F731N');
    expect(quirk.resolution).toBe(HingeResolution.Detents);
    expect(quirk.detentValues).toEqual([0, 90, 180]);
  });

  it('matches case-insensitively', () => {
    expect(lookupQuirk('SAMSUNG', 'sm-f731n').resolution).toBe(
      HingeResolution.Detents,
    );
  });

  it('lets a runtime registration override a shipped entry', () => {
    // An app must be able to correct a bad entry without waiting for us.
    registerQuirk('samsung/sm-f731n', {
      ...ASSUMED_QUIRK,
      resolution: HingeResolution.Continuous,
    });
    expect(lookupQuirk('samsung', 'SM-F731N').resolution).toBe(
      HingeResolution.Continuous,
    );
  });

  it('falls back to a manufacturer-wide wildcard', () => {
    registerQuirk('weirdcorp/*', { ...ASSUMED_QUIRK, inverted: true });
    expect(lookupQuirk('weirdcorp', 'wc-99').inverted).toBe(true);
  });

  it('is safe with a null manufacturer', () => {
    expect(lookupQuirk(null, null)).toBe(ASSUMED_QUIRK);
  });
});

File: flutter/test/quirks_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:hinge_devices/hinge_devices.dart';

void main() {
  setUp(FoldableQuirks.reset);

  group('normalisation', () {
    test('passes through a standard 0-180 device', () {
      const n = HingeNormalizer(HingeQuirk(range: HingeAngleRange.zeroTo180));
      expect(n.normalize(90), 90);
      expect(n.normalize(0), 0);
      expect(n.normalize(180), 180);
    });

    test('null in, null out — a missing sensor never becomes zero', () {
      const n = HingeNormalizer(HingeQuirk.assumed);
      expect(n.normalize(null), isNull);
    });

    test('mirrors a 0-360 device past the flat point', () {
      const n = HingeNormalizer(HingeQuirk(range: HingeAngleRange.zeroTo360));
      expect(n.normalize(180), 180);
      expect(n.normalize(270), 90);
      expect(n.normalize(360), 0);
    });

    test('inverts a device that reports 180 closed', () {
      const n = HingeNormalizer(
        HingeQuirk(range: HingeAngleRange.zeroTo180, inverted: true),
      );
      expect(n.normalize(0), 180);
      expect(n.normalize(180), 0);
    });

    test('applies an offset before normalising', () {
      const n = HingeNormalizer(
        HingeQuirk(range: HingeAngleRange.zeroTo180, offset: 5),
      );
      expect(n.normalize(85), 90);
    });

    test('clamps out-of-range readings', () {
      const n = HingeNormalizer(HingeQuirk.assumed);
      expect(n.normalize(-10), 0);
      expect(n.normalize(200), 180);
    });
  });

  group('lookup', () {
    test('falls back to the assumed convention for unknown devices', () {
      expect(
        FoldableQuirks.lookup('nothing', 'phone-3').range,
        HingeAngleRange.zeroTo180,
      );
    });

    test('matches an exact device key, case-insensitively', () {
      FoldableQuirks.register(
        'weirdcorp/wc-1',
        const HingeQuirk(range: HingeAngleRange.zeroTo360),
      );
      expect(
        FoldableQuirks.lookup('WeirdCorp', 'WC-1').range,
        HingeAngleRange.zeroTo360,
      );
    });

    test('falls back to a manufacturer-wide wildcard', () {
      FoldableQuirks.register(
        'weirdcorp/*',
        const HingeQuirk(range: HingeAngleRange.zeroTo180, inverted: true),
      );
      expect(FoldableQuirks.lookup('weirdcorp', 'wc-99').inverted, isTrue);
    });

    test('an exact entry beats the wildcard', () {
      FoldableQuirks.register(
        'weirdcorp/*',
        const HingeQuirk(range: HingeAngleRange.zeroTo180, inverted: true),
      );
      FoldableQuirks.register(
        'weirdcorp/wc-1',
        const HingeQuirk(range: HingeAngleRange.zeroTo180),
      );
      expect(FoldableQuirks.lookup('weirdcorp', 'wc-1').inverted, isFalse);
    });

    test('a null manufacturer is safe', () {
      expect(
          FoldableQuirks.lookup(null, null).range, HingeAngleRange.zeroTo180);
    });

    test('shipped entries are found without registration', () {
      // Measured on hardware: the Z Flip 5 reports exactly three values.
      final quirk = FoldableQuirks.lookup('samsung', 'SM-F731N');
      expect(quirk.resolution, HingeResolution.detents);
      expect(quirk.detentValues, <double>[0, 90, 180]);
    });

    test('a runtime registration overrides a shipped entry', () {
      // An app must be able to correct a bad entry without waiting for us.
      FoldableQuirks.register(
        'samsung/sm-f731n',
        const HingeQuirk(
          range: HingeAngleRange.zeroTo180,
          resolution: HingeResolution.continuous,
        ),
      );
      expect(
        FoldableQuirks.lookup('samsung', 'SM-F731N').resolution,
        HingeResolution.continuous,
      );
    });

    test('an unknown device does not claim to be continuous', () {
      // Assuming a smooth sweep and being wrong produces an effect that
      // visibly snaps, which is worse than not offering the effect.
      expect(
        FoldableQuirks.lookup('nothing', 'phone-3').resolution,
        HingeResolution.unknown,
      );
    });
  });

  group('resolution guides effect decisions', () {
    test('only a continuous hinge supports continuous effects', () {
      const continuous = Hinge(resolution: HingeResolution.continuous);
      const detents = Hinge(resolution: HingeResolution.detents);
      const unknown = Hinge();

      expect(continuous.supportsContinuousEffects, isTrue);
      expect(detents.supportsContinuousEffects, isFalse);
      expect(
        unknown.supportsContinuousEffects,
        isFalse,
        reason: 'unknown must not be optimistic',
      );
    });
  });
}
